File: backend/debug.py
"""
Debug utility module for CoolChat backend.

Loads and manages debug configurations from debug.json file.
"""

import json
import os
from typing import Dict, Any


class DebugLogger:
    def __init__(self):
        self.config: Dict[str, Any] = {}
        self.load_config()

    def load_config(self):
        """Load debug configuration from debug.json file."""
        debug_file = os.path.join(os.path.dirname(__file__), "..", "debug.json")
        try:
            with open(debug_file, 'r') as f:
                self.config = json.load(f)
        except Exception as e:
            print(f"Warning: Failed to load debug.json: {e}")
            # Default config if file can't be loaded
            self.config = {
                "backend": {
                    "tool_calls": True,
                    "llm_requests": False,
                    "llm_responses": False,
                    "database_operations": False,
                    "file_operations": False,
                    "performance": False
                },
                "enabled": True
            }

    def is_enabled(self, category: str) -> bool:
        """Check if a specific debug category is enabled."""
        if not self.config.get("enabled", True):
            return False

        backend_config = self.config.get("backend", {})
        return backend_config.get(category, False)

    def log(self, category: str, message: str, *args):
        """Log a message if the category is enabled."""
        if self.is_enabled(category):
            print(f"[DEBUG {category.upper()}] {message}", *args)

    def debug_tool_calls(self, message: str):
        """Log tool call debugging."""
        self.log("tool_calls", message)

    def debug_llm_requests(self, message: str):
        """Log LLM request debugging."""
        self.log("llm_requests", message)

    def debug_llm_responses(self, message: str):
        """Log LLM response debugging."""
        self.log("llm_responses", message)

    def debug_db(self, message: str):
        """Log database operation debugging."""
        self.log("database_operations", message)

    def debug_files(self, message: str):
        """Log file operation debugging."""
        self.log("file_operations", message)

    def debug_perf(self, message: str):
        """Log performance debugging."""
        self.log("performance", message)


# Global debug logger instance
_debug_logger = None

def get_debug_logger() -> DebugLogger:
    """Get the global debug logger instance."""
    global _debug_logger
    if _debug_logger is None:
        _debug_logger = DebugLogger()
    return _debug_logger

def debug_log(category: str, message: str, *args):
    """Convenience function for debug logging."""
    logger = get_debug_logger()
    logger.log(category, message, *args)

File: frontend/src/debug.js
/**
 * Debug utility module for CoolChat frontend.
 *
 * Loads and manages debug configurations from debug.json file.
 */
import { API_BASE } from './api.js';

class DebugLogger {
  constructor() {
    this.config = null;
    this.lastConfigCheck = 0;
    this.configCacheTime = 5000; // 5 seconds
    this.loadConfig();
  }

  // For frontend: Fetch debug config from root directory
  async loadConfig() {
    const now = Date.now();

    // Use cached config if recent enough
    if (this.config && now - this.lastConfigCheck < this.configCacheTime) {
      return await this.config;
    }

    try {
      const response = await fetch(`${API_BASE}/debug.json`);
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      this.config = await response.json();
      this.lastConfigCheck = now;
    } catch (error) {
      // Warning to avoid spam
      if (!this.config) {
        console.warn('Failed to load debug.json - using defaults');
      }
      // Default config if file can't be loaded
      this.config = {
        "frontend": {
          "tool_parsing": true,
          "llm_responses": true,
          "api_calls": false,
          "store_actions": false,
          "component_render": false,
          "performance": false
        },
        "enabled": true
      };
    }

    return this.config;
  }

  async isEnabled(category) {
    const config = await this.loadConfig();
    if (!config || !config.enabled) {
      return false;
    }

    const frontendConfig = config.frontend || {};
    return frontendConfig[category] === true;
  }

  async log(category, message, ...args) {
    if (await this.isEnabled(category)) {
      console.log(`[DEBUG ${category.toUpperCase()}] ${message}`, ...args);
    }
  }

  async debugToolParsing(message, ...args) {
    await this.log('tool_parsing', message, ...args);
  }

  async debugLLMResponses(message, ...args) {
    await this.log('llm_responses', message, ...args);
  }

  async debugAPICalls(message, ...args) {
    await this.log('api_calls', message, ...args);
  }

  async debugStoreActions(message, ...args) {
    await this.log('store_actions', message, ...args);
  }

  async debugComponentRender(message, ...args) {
    await this.log('component_render', message, ...args);
  }

  async debugPerformance(message, ...args) {
    await this.log('performance', message, ...args);
  }
}

// Global instance
let _debug_logger = null;

export function getDebugLogger() {
  if (!_debug_logger) {
    _debug_logger = new DebugLogger();
  }
  return _debug_logger;
}

export function debugLog(category, message, ...args) {
  const logger = getDebugLogger();
  return logger.log(category, message, ...args);
}

// Convenience functions
export const debugToolParsing = (message, ...args) => {
  const logger = getDebugLogger();
  return logger.debugToolParsing(message, ...args);
};

export const debugLLMResponses = (message, ...args) => {
  const logger = getDebugLogger();
  return logger.debugLLMResponses(message, ...args);
};

export const debugAPICalls = (message, ...args) => {
  const logger = getDebugLogger();
  return logger.debugAPICalls(message, ...args);
};

export const debugStoreActions = (message, ...args) => {
  const logger = getDebugLogger();
  return logger.debugStoreActions(message, ...args);
};

export const debugComponentRender = (message, ...args) => {
  const logger = getDebugLogger();
  return logger.debugComponentRender(message, ...args);
};

export const debugPerformance = (message, ...args) => {
  const logger = getDebugLogger();
  return logger.debugPerformance(message, ...args);
};